### DISCUSSION TITLE
Does IFC preserve information about type decorations?

### DISCUSSION THREAD

**Discussion by AndreyG** [microsoft/ifc-spec#7] [General]
Let's consider function `say_hello` declared in a module interface unit in the following way
```
void say_hello(std::ostream&, const std::string_view& name);
```
Does IFC provide a way to restore this pretty type for the function? The only result I managed to achieve is
```
void (basic_ostream<char, char_traits<char>>&, const basic_string_view<char, char_traits<char>>&)
```

**Comment by AndreyG** [comment]
Сircumstantial evidence of my assumption is the following experiment. I tried to invoke `say_hello` with wrong arguments, i.e. 
```
say_hello(1, "abacaba)"
```
and compiled this code by MSVC. When `say_hello` is defined in the same translation unit as the invocation the error is
```
... cannot convert argument 1 from 'int' to 'std::ostream &'
```
but when `say_hello` is imported from another unit, the error turns into
```
... cannot convert argument 1 from 'int' to 'basic_ostream<char, char_traits<char>> &'
```

**Comment by GabrielDosReis** [comment]
> Does IFC provide a way to restore this pretty type for the function?

What you get from the IFC is what the compiler put there, so the IFC itself does not have a mechanism to put the type alias there.  It is a gap in the way that the compiler (MSVC in this case) writes out the IFC - which is focused on the semantics - as most type aliases would have been evaluated by the time the IFC is written out to disk.  This is also related to the issue GabrielDosReis/ipr#26  regarding how to handle type aliases.  See also discussion in section 4.3 of the [IPR paper](https://www.stroustrup.com/macis09.pdf)

**Reply by AndreyG** [reply]
Many thanks for the quick reply and for the references!

Do I understand correctly, that plan is to fix loosing of type decorations in MSVC and then reflect this change in IFC? IMHO the fact that transition from header files to modules dramatically reduces quality of compiler messages (compare the first and second error messages from my [previous comment](https://github.com/microsoft/ifc-spec/discussions/7#discussioncomment-1556843)) is a stopper for wide adoption of modules.

Moreover it's impossible to consider modules as a "Tooling Opportunity" while this issue isn't fixed, because all tools -- IDE/static analyzer/etc... -- need to interact with the user providing him messages referencing code, and consequently `basic_string<char, char_traits<char>, allocator<char>>` instead of `string` is not acceptable.

**Comment by GabrielDosReis** [comment]
> Do I understand correctly, that plan is to fix loosing of type decorations in MSVC and then reflect this change in IFC?

The compiler does not lose any type information.  `std::ostream` is a type alias for the expanded version of the `basic_stream` template specialization.  So the codegen will still be the same.  What indeed is lost is the retention of the type alias in the diagnostics -- this is an issue in the MSVC compiler itself (and the code for recovering the type alias from its expansion is quite complicated and does not always work beyond trivial cases like this, independently of modules).  That is an issue that the MSVC compiler team is working on, from a more principled representation that avoids all these issues.

>  IMHO the fact that transition from header files to modules dramatically reduces quality of compiler messages (compare the first and second error messages from my previous comment) is a stopper for wide adoption of modules.

That is one view, more pessimistic than I would want to bet on.  I am more optimistic that modules are actually bringing implementers to work more principled diagnostic machinery.  I would encourage you to continue to experiment; that is how we make progress and adoption at large.

> Moreover it's impossible to consider modules as a "Tooling Opportunity" while this issue isn't fixed, because all tools

That is certainly a dramatic statement and view that I don't share: we wouldn't be having this discussion if we didn't have Modules in the first place, and it is an example of tooling opportunity :-)

**Comment by C-EDG** [comment]
I have to say, I largely agree with Andrey that the loss of type alias information is a concern. I believe the IFC spec allows for this information to be encoded and therefore it's less an issue with IFC (assuming I have it correct) and more an issue with what MSVC puts into IFC.

My understanding is that one of the goals of IFC is to be a universal, portable spec. Assuming that's still a goal, I think it's very worthwhile to discuss IFC independently of what MSVC is doing with it, and to separate those concerns.

Which is to say that, while MSVC currently is the reference compiler for IFC, it's worthwhile to separate issues between "MSVC doesn't do this" and "IFC doesn't enable this".

Similarly, it's important to keep in mind that MSVC's way of doing things isn't the only way, so just because something works for MSVC doesn't mean that it's overly feasible for all consumers of IFC to do things the same way - so it's important to have eyes with different views and experiences on this project and contributing their thoughts, concerns, ideas, etc., and give those thoughts due consideration.

I think it's also well worth keeping in mind that IFC is in its infancy, and recognize that it started with MSVC and so it's naturally going to align quite well with how MSVC does things. The next goal for IFC is to expand its usefulness so that others can see good reason to adopt it - and that's up to those of us who are expending effort on IFC to work towards :)

**Reply by GabrielDosReis** [reply]
> I believe the IFC spec allows for this information to be encoded and therefore it's less an issue with IFC (assuming I have it correct) and more an issue with what MSVC puts into IFC.

Exactly right.

> My understanding is that one of the goals of IFC is to be a universal, portable spec. Assuming that's still a goal, I think it's very worthwhile to discuss IFC independently of what MSVC is doing with it, and to separate those concerns.

Again, agreed.

> Which is to say that, while MSVC currently is the reference compiler for IFC, it's worthwhile to separate issues between "MSVC doesn't do this" and "IFC doesn't enable this".

Agreed.

> Similarly, it's important to keep in mind that MSVC's way of doing things isn't the only way, so just because something works for MSVC doesn't mean that it's overly feasible for all consumers of IFC to do things the same way - so it's important to have eyes with different views and experiences on this project and contributing their thoughts, concerns, ideas, etc., and give those thoughts due consideration.

We are in violent agreement.

> I think it's also well worth keeping in mind that IFC is in its infancy, and recognize that it started with MSVC and so it's naturally going to align quite well with how MSVC does things. The next goal for IFC is to expand its usefulness so that others can see good reason to adopt it - and that's up to those of us who are expending effort on IFC to work towards :)

In my CppCon2021 talk, I explicitly call out need for “gradual lifting”, i.e. abstracting away from implementation details to achieve the goals set out for the IFC.  That would use input from people like you and @AndreyG
